Step3 구간 추가 기능 리뷰 부탁드립니다!

ilmare-cbk opened the pull request:
안녕하세요 우테캠 Pro 5기 교육생 천보경입니다.
이번 미션은 먼저 인수조건을 작성하고,
인수조건을 만족하기 위해 테스트를 작성하고 기능을 구현하는 방향으로 개발하고자 신경썼습니다!

리뷰 잘 부탁드립니다!

ilmare-cbk committed:
docs(README.md): step3 요구사항 정리

ilmare-cbk committed:
test(subway package): 역 사이에 새로운 역 등록 인수 테스트 작성

- Section 엔티티 구현
- Section 과 Line 다대일 양방향 연관관계 매핑
- Section, Sections 도메인 모델에 대한 테스트 추가

ilmare-cbk committed:
test(SectionAcceptanceTest.java): 새로운 역을 상행 종점으로 등록하는 인수 테스트 추가

ilmare-cbk committed:
test(SectionAcceptanceTest.java): 새로운 역을 하행 종점으로 등록하는 인수 테스트 추가

ilmare-cbk committed:
test(SectionAcceptanceTest.java): 구간 등록 예외 케이스 인수 테스트 추가

- 역 사이에 새로운 역을 등록할 경우 기존 역 사이 길이보다 크거나 같으면 등록할 수 없다.

ilmare-cbk committed:
test(SectionAcceptanceTest.java): 구간 등록 예외 케이스 인수 테스트 추가

- 상행역과 하행역이 이미 노선에 모두 등록되어 있다면 추가할 수 없다.

ilmare-cbk committed:
test(SectionAcceptanceTest.java): 구간 등록 예외 케이스 인수 테스트 추가

- 상행역과 하행역 둘 중 하나도 포함되어있지 않으면 추가할 수 없다.

dhmin5693 commented on src/main/java/nextstep/subway/section/domain/Section.java:
```diff
@@ -0,0 +1,110 @@
+    @JoinColumn(name = "line_id")
+    private Line line;
+
+    private int distance;
```
`Distance` 도 별도의 책임을 부여할 수 있겠네요.

ilmare-cbk replied:
네 distance 뿐만 아니라 Line, Station의  name과 Line의 color도 wrapping된 객체로 책임을 부여하도록 수정해보겠습니다

dhmin5693 replied:
좋습니다!

dhmin5693 commented on src/test/java/nextstep/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,168 @@
+
+@DisplayName("구간 관련 기능")
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class SectionAcceptanceTest {
```
이 클래스의 모든 테스트가 실패하고 있습니다. 다시 한번 확인해주세요.

ilmare-cbk replied:
@dhmin5693 
동휘님 step3 브랜치에서 테스트 돌렸을 때 저는 다 성공하고 있어서요
혹시 다시 한 번 확인 부탁드려도 될까요?? 😃

ilmare-cbk replied:
build and run 작동 시 intellij idea로 변경하여 실행해보니 테스트 실패하는 것 확인했습니다!

dhmin5693 commented on src/main/java/nextstep/subway/common/message/ExceptionMessage.java:
```diff
@@ -4,7 +4,10 @@ public class ExceptionMessage {
-    public static final String UP_STATION_EQUALS_DOWN_STATION = "상행좀점역과 하행종점역은 같을 수 없습니다.";
+    public static final String UP_STATION_EQUALS_DOWN_STATION = "상행역과 하행역은 같을 수 없습니다.";
+    public static final String INVALID_SECTION_DISTANCE = "지하철 구간의 길이는 0보다 큰 값이어야 합니다. (현재 = %d)";
+    public static final String ALREADY_ADDED_SECTION = "이미 등록된 자하철 구간입니다.";
```
```suggestion
    public static final String ALREADY_ADDED_SECTION = "이미 등록된 지하철 구간입니다.";
```
오타가 끼어 있네요

ilmare-cbk replied:
넵! 오타 수정하겠습니다!

dhmin5693 commented on src/main/java/nextstep/subway/section/domain/Sections.java:
```diff
@@ -0,0 +1,91 @@
+        return results;
+    }
+
+    private void addDownStationsOfSection(Optional<Section> preSection, List<Station> results) {
```
collection을 매개변수로 받은 뒤 그 내부에 넣어주는 방식은 유지보수가 굉장히 어려운 편입니다.
`Map<Object, Object>` 가 수많은 메소드를 매개변수로 넣어주고 그 안에서 데이터를 넣고 있다고 생각해보세요.
어디서 어떤 데이터가 들어갔는지 찾기 어려울 것 같지 않나요?

데이터를 반환한 다음 호출한 곳에서 합치는 방식을 추천드립니다.

ilmare-cbk replied:
네 이부분 켈렉션을 파라미터로 넘기지 않도록 수정했습니다!

dhmin5693 commented on src/main/java/nextstep/subway/section/domain/Sections.java:
```diff
@@ -0,0 +1,91 @@
+    public List<Station> getStationsInOrder() {
+        Optional<Section> preSection = findSectionIncludingUpStation();
+
+        List<Station> results = Lists.newArrayList();
```
addDownStationsOfSection가 개선되면 여기서 미리 가변 리스트를 선언하지 않고 바로 불변의 리스트를 만들 수 있을 것 같아요.

ilmare-cbk replied:
리팩토링하면서 해당 메소드는 제거되었습니다!

dhmin5693 commented on src/main/java/nextstep/subway/section/domain/Section.java:
```diff
@@ -0,0 +1,110 @@
+        validateStationIsNotNull(upStation);
+        validateStationIsNotNull(downStation);
+
+        if (upStation.equals(downStation)) {
```
station의 equals가 name만 비교하던데 id는 안보아도 괜찮을까요?
얼핏 생각하면 name이 unique 조건이라 상관 없을듯 하지만 id까지 검사한다면 영속성 여부까지 확인할 수 있습니다.

또 이번 미션에 포함된 스펙은 아니지만 현실에선 하나의 정거장이 여러 노선에 포함된 케이스가 있죠.
그런 경우를 다 확인해보려면 ID도 비교해주는게 좋다고 생각합니다.

ilmare-cbk replied:
말씀해주신대로 id값도 같이 비교하도록 수정했습니다

dhmin5693 commented on src/test/java/nextstep/subway/section/domain/SectionTest.java:
```diff
@@ -0,0 +1,91 @@
+        this.distance = 10;
+    }
+
+    @DisplayName("동등성 실패")
```
동등성이 실패하는 조건은 많습니다.
어떤 조건이 아니면 실패하는지를 정확히 기입해주시는건 어떤가요?

보통 미지의 코드를 가장 먼저 보아야 한다면 테스트 코드를 보는 경우가 있습니다.
코드가 어떤 조건에서 어떻게 동작되어야 하는지를 잘 기술해두었기 때문이에요.
다른 말로 하면 docs 역할도 겸하고 있는 것이죠.

그런 관점에서 `DisplayName` 을 수정해보시기 바라요.

ilmare-cbk replied:
네 테스트 코드가 docs 역할을 한다는 점을 봤을 때
조금더 명확하고 정확하게 작성해야 될 것 같네요
말씀해주신 부분들 수정해보겠습니다!

dhmin5693 commented on src/test/java/nextstep/subway/section/domain/SectionsTest.java:
```diff
@@ -0,0 +1,84 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class SectionsTest {
```
케이스 꼼꼼하네요 💯

dhmin5693 commented:
커밋 단위를 잘 잘라서 메시지가 굉장히 깔끔하게 작성된 점이 굉장히 눈에 띄네요. 💯 
여기서 첨언을 드리자면 테스트는 feature의 검증을 위한 부가적인 부분이에요.
feature가 한 커밋에 포함되었다면 feature의 우선순위가 가장 높은게 좋습니다.

주된 요소와 부가 요소 중 무엇을 표현할까의 차이라고 봐주시면 되겠습니다.

dhmin5693 reviewed (changes requested):
보경님 안녕하세요!

늦은 새벽에 요청주셨던데, 정말 대단하신 열정이네요.
하지만 아직 일정이 꽤 많이 남았으니 적당히 페이스를 유지하는게 좋습니다.
너무 빠르게 달리면 쉽게 지치더라구요 🥲

바로 머지를 누르고 싶지만 실패하는 테스트가 있으니 한번 더 확인이 필요합니다.
다른 의견도 함께 검토 부탁드릴게요.

ilmare-cbk commented:
> 커밋 단위를 잘 잘라서 메시지가 굉장히 깔끔하게 작성된 점이 굉장히 눈에 띄네요. 💯 여기서 첨언을 드리자면 테스트는 feature의 검증을 위한 부가적인 부분이에요. feature가 한 커밋에 포함되었다면 feature의 우선순위가 가장 높은게 좋습니다.
> 
> 주된 요소와 부가 요소 중 무엇을 표현할까의 차이라고 봐주시면 되겠습니다.

리뷰 감사합니다!
이후에는 feature와 test가 한 커밋에 포함되는 경우 feature의 우선순위를 높여 커밋 메세지를 작성하도록 해보겠습니다!

ilmare-cbk committed:
feat(Distance.java): Section의 distance(int) 를 Distance 클래스로 wrapping

- Distance 객체에서 거리 값 유효성 검사를 하도록 수정
- Distance 객체의 테스트 코드 추가

ilmare-cbk committed:
feat(SubwayValidator.java): 지하철 노선 이름과 지하철역 이름 유효성 검사하는 클래스 생성

- 지하철 노선 이름과 지하철 이름 유효성 검사 로직이 같기 때문에 SubwayValidator 에서 검사하도록 분리
- SubwayValidator 클래스 테스트 추가

ilmare-cbk committed:
feat(LineName.java): 지하철 노선 이름을 LineName 클래스로 wrapping

- LineName 객체에서 지하철 노선 이름의 유효성 검사를 하도록 수정
- LineName 객체의 테스트 코드 추가

ilmare-cbk committed:
feat(LineColor.java): 지하철 노선 색상을 LineColor 클래스로 wrapping

- LineColor 객체에서 지하철 노선 색상의 유효성 검사를 하도록 수정
- LineColor 객체의 테스트 코드 추가

ilmare-cbk committed:
feat(StationName.java): 지하철역 이름을 StationName 클래스로 wrapping

- StationName 객체에서 지하철역 이름의 유효성 검사를 하도록 수정
- StationName 객체의 테스트 코드 추가

ilmare-cbk committed:
refactor(ExceptionMessage.java): 오타 수정

ilmare-cbk committed:
refactor(dto package): SectionAcceptanceTest 실패하는 테스트 수정

- 기본 생성자 추가해서 deserialize 되도록 수정

ilmare-cbk committed:
refactor(Station.java): station 동등성 비교 시 id 필드도 포함되도록 수정

ilmare-cbk committed:
refactor(Sections.java): collection을 메소드 파라미터로 넘기지 않도록 수정

ilmare-cbk committed:
refactor(Sections.java): getDistance 메소드 호출 시 Distance 객체를 리턴하도록 수정

- Distance 클래스의 subtract 메소드 추가 및 테스트 코드 추가

ilmare-cbk committed:
refactor(test package): 테스트가 명확한 의미를 나타내도록 displayName 변경

dhmin5693 reviewed (approved):
훌륭하시네요!
마지막 미션인 4단계 시작하셔도 될 것 같습니다.
마무리해주세요 :)
수고하셨습니다.
